debug/org.eclipse.cdt.debug.ui/src/org/eclipse/cdt/debug/internal/ui/commands/ToggleInstructionStepModeHandler.java

```diff
@@ -0,0 +1,186 @@
+				boolean currentState = (Boolean) state.getValue();
+				if (currentState != prefState) {
+					HandlerUtil.toggleCommandState(command);
+				}
+			} catch (ExecutionException e) {
+				CDebugUIPlugin.log(e);
+			}
+		}
+	}
+
+	@Override
+	public void setEnabled(Object evaluationContext) {
+		ICommandService commandService = PlatformUI.getWorkbench().getService(ICommandService.class);
+		if (commandService != null) {
+			this.command = commandService.getCommand(TISM_COMMAND_ID);
+
+		}
+	}
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException {
+		State state = command.getState(RegistryToggleState.STATE_ID);
+		if (state == null) {
+			throw new ExecutionException(NLS.bind(Messages.ToogleCommand_State_Not_found, command.getId()));
+		}
+
+		boolean currentState = (Boolean) state.getValue();
+		HandlerUtil.toggleCommandState(command);
+
+		ISteppingModeTarget target = CDTDebugPropertyTester.getSteppingModeTarget();
+		if (target != null) {
+			target.enableInstructionStepping(!currentState);
+		}
+
+		return IStatus.OK;
+	}
+
+	private boolean isTerminated(ISteppingModeTarget target) {
+		return ((target instanceof ITerminate && ((ITerminate) target).isTerminated())
+				|| (target instanceof IDisconnect && ((IDisconnect) target).isDisconnected()));
+	}
+
+	@Override
+	public void dispose() {
+		if (workbenchWindow != null) {
+			DebugUITools.getDebugContextManager().getContextService(workbenchWindow).removeDebugContextListener(this);
+		}
+		ISteppingModeTarget target = getTarget();
+		if (target != null && target instanceof ITargetProperties) {
+			((ITargetProperties) target).removePropertyChangeListener(this);
+		}
+		setTarget(null);
+		super.dispose();
+	}
+
+	@Override
+	public void debugContextChanged(DebugContextEvent event) {
+		ISelection selection = event.getContext();
+		ISteppingModeTarget newTarget = null;
+		if (selection instanceof IStructuredSelection) {
+			newTarget = getTargetFromSelection(((IStructuredSelection) selection).getFirstElement());
+		}
+
+		if (newTarget == null) {
+			return;
+		}
+
+		ISteppingModeTarget oldTarget = getTarget();
+		if (newTarget.equals(oldTarget)) {
+			return;
+		}
+
+		setTarget(newTarget);
+		if (newTarget instanceof ITargetProperties) {
+			((ITargetProperties) newTarget).addPropertyChangeListener(this);
+		}
+
+		try {
+			boolean prefState = !isTerminated(newTarget) && newTarget.isInstructionSteppingEnabled();
+			if (command != null) {
+				State state = command.getState(RegistryToggleState.STATE_ID);
+				boolean currentState = (Boolean) state.getValue();
+				if (currentState != prefState) {
+					HandlerUtil.toggleCommandState(command);
+				}
+			}
+		} catch (ExecutionException e) {
+			CDebugUIPlugin.log(e);
+		}
+	}
+
+	private ISteppingModeTarget getTarget() {
+		return this.fTarget;
+	}
+
+	private void setTarget(ISteppingModeTarget target) {
+		this.fTarget = target;
+	}
+
+	private ISteppingModeTarget getTargetFromSelection(Object element) {
```

Comment:
Can we avoid duplicating the last part of `CDTDebugPropertyTester.getSteppingModeTarget()` and reuse it here instead?

Reply:
Fixed.